### PR TITLE
kwil-admin docs

### PR DESCRIPTION
**Pull request by jchappelow**
Initial `kwil-admin` docs.  Lots in the `node` command page is premature and likely to change since those features are not yet merged.

image

The Installation section is best guess at this point since we should probably do two things:
- publish the kwild:latest as the images created from the kwild.script.dockerfile, which will allow entering a command prompt in the docker container
- include `kwil-admin` in the binary release archives published on https://github.com/kwilteam/kwil-cli/releases, at least until we open source the kwil-db repo

### CODE REVIEW

File: docs/admin/install.mdx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+With a running container (e.g. `docker container run kwild:latest`), open an
+interactive shell in the container where you can run `kwil-admin` commands:
+
+```sh
+docker exec -it <container-name-or-id> /bin/sh
+```
+
+**DRAFT NOTE: this only works with the "shell" variant of the docker image, which is ~35MiB larger than the scratch one.  Maybe we make that standard?**
+
+## Download the standalone `kwil-admin` executable
```
Comment (jchappelow) [inline docs/admin/install.mdx:31]: Need title case consistency

File: docs/admin/index.mdx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+1. Quick test network configuration.
+2. Node key generation.
+3. Interaction with a running `kwild` process via the node's authenticated gRPC interface.
+4. Validator operations.
+
+```mdx-code-block
+import DocCardList from '@theme/DocCardList';
+
+<DocCardList />
+```
```
Comment (jchappelow) [inline docs/admin/index.mdx:22]: Doing this instead of auto-generated index page via _category_.json file so we can have actual page content above it.

File: docs/admin/node.mdx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+The node will generate its own TLS key and certificate, which is used encrypt
+the communications on its authenticated RPC server. As a convenience, `kwild`
+will also generate new *client* credentials on start up if the file
+`~/.kwild/clients.pem` does not exist.
+
+The corresponding `auth.{cert,key}` files may then be used by the operator to
+with th `kwil-admin node` commands for remote access to the node. 
+
+## Generation of Credentials by `kwil-admin`
```
Comment (jchappelow) [inline docs/admin/node.mdx:64]: I should clarify that these are orthogonal generation approaches.  The first is convenient for testing and avoids the need to restart the node.  The second would be helpful for making a set of credentials that an operator of multiple nodes imports to all of their nodes.

File: docs/admin/node.mdx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+node's `clients.pem` file. Specifically,
+
+1. Generate credentials with `./kwil-admin node gen-auth-key`
+2. Securely copy the `~/.kwil-admin/auth.cert` file to the machine where `kwild`
+   is running, if it is not the same host.
+3. Append the contents of the copied `auth.cert` to the node's `clients.pem` file.
+
+  ```sh
+  cat auth.cert >> ~/.kwild/clients.pem
+  ```
```
Comment (jchappelow) [inline docs/admin/node.mdx:79]: This process may seen a bit complex at first, but it's just a variation of pre-shared key scheme.  Say we had a (less secure and more complex) password or token approach, then we'd be editing the node config with say `authrpcpass=adsadsf` and passing that same password to `kwil-admin` (and similar for a token).

File: docs/admin/node.mdx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  }
+}
+```
+
+### Peers
+
+A list of the peers currently connected to the node are reported with the
+`peers` subcommand.
+
+... TODO EXAMPLE output
```
Comment (jchappelow) [inline docs/admin/node.mdx:160]: I actually haven't run the peers subcommand on a testnet with >1 node.  Right now I just get back `[]` for my single node network.  :)

File: docs/admin/setup.mdx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+the files in the default node root directory at `~/.kwild`.
+
+The paths to the SQLite files (`--sqlpath`) and snapshots (`--snappath`) are
+optional, but provided here since they are configurable on the node.
+
+## Resetting Only Chain state
+
+To reset the block chain state data while leaving all of the other application
+data such as the SQLite databases intact, use the `reset-state` command. This
+will delete the data files under `<root_dir>/abci/data`
```
Comment (jchappelow) [inline docs/admin/setup.mdx:130]: @charithabandi I actually wasn't terribly clear on the use case for `reset-state`.  Would you add or change the description here?

File: docs/admin/validators.mdx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+The `validators` command provides functions for creating and broadcasting
+validator-related transactions (join/approve/leave), and retrieving information
+on the current validators and join requests.
+
+```
+kwil-admin validators [--rpcserver RPCSERVER] <command> [<args>]
+```
+
+Each of the subcommands uses a Kwil node's public RPC service (`--rpcserver`),
+where the default is "127.0.0.1:50051".
```
Comment (jchappelow) [inline docs/admin/validators.mdx:19]: As such, I suspect we'll either see this command go away or become a node subcommand if it starts using the authenticated RPC service to perform server-side tx auth/signing.

File: docs/admin/setup.mdx (hunk 6 above)
Comment (charithabandi) [reply docs/admin/setup.mdx:130]: This is more for the development purposes and not to be used in the production. 

It basically resets the chain state, that means it deletes any data generated by the wal like mempool wal, block data, state data etc.  So everything under `<root_dir>/abci/data` and `<root_dir>/abci/info`

Comment (jchappelow) [reply docs/admin/setup.mdx:130]: What I'm getting at is the specific development workflows where `reset-state` would be used.  Do you use it regularly?  If you delete state but not the rest of the application data, what then?  Is that where you'd use the `genesis-hash` subcommand?

I'm a bit hesitant to put commands in kwil-admin that are for development only, particularly if I don't see the workflow that would demand it.

Comment (charithabandi) [reply docs/admin/setup.mdx:130]: Yeah, reset-state can be removed, we use reset-all for more of our testing

**Comment by KwilLuke** [comment]
Should the Kwil Admin Tool be contained within Kwil Daemon or its own separate section?

My initial instinct would be within Kwil Daemon, but I am curious the logic for its own section.

**Comment by jchappelow** [comment]
> Should the Kwil Admin Tool be contained within Kwil Daemon or its own separate section?
> 
> My initial instinct would be within Kwil Daemon, but I am curious the logic for its own section.

Nesting it should be fine.  Easy to move.  No strong reasoning other than the large scope of the section with multiple subsections.

File: docs/admin/index.mdx (hunk 2 above)
Comment (brennanjl) [reply docs/admin/index.mdx:22]: I like this a lot.  We should adopt this across more of the docs.

File: docs/admin/setup.mdx [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+```
+kwil-admin setup testnet [--validators V] [--non-validators N] [--config FILE]
+  [--output-dir DIR] [--node-dir-prefix PRE] [--hostname-prefix PRE]
+  [--hostname-suffix SUF] [--starting-ip IP] [--hostname HOSTNAME] [--p2p-port PORT]
+```
+
+There are defaults for each flag, with the output writing to a folder called
+`.testnet` in the current folder.  For example, the following command creates a
+new testnet work with 8 total nodes comprising 4 validators and 4 non-validators:
```
Comment (brennanjl) [inline docs/admin/setup.mdx:54]: typo "testnet work", instead of "test network"

File: docs/admin/setup.mdx [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+hash" in `kwild`'s genesis configuration to reflect the initial state that
+includes one or more SQLite files. The syntax is:
+
+```
+kwil-admin setup genesis-hash [--genesis GENESIS] [DBDIR]
+```
+
+Both inputs are optional, using default paths. `GENESIS` is the path to the
+genesis file to patch, which is `~/.kwild/abci/config/genesis.json` by default.
+The `DBDIR` argument is the directory containing all of kwild's .sqlite files to
```
Comment (brennanjl) [inline docs/admin/setup.mdx:105]: I think it should maybe be `kwild`'s.

Also, should it be `.sqlite`?  Not sure

File: docs/admin/validators.mdx (hunk 7 above)
Comment (brennanjl) [reply docs/admin/validators.mdx:19]: Kind've a weird one.

I think `join-status` should probably be on the public endpoint (however if this makes config weird / confusing by having to specify 2 endpoints, it's probably fine to include it with the rest).

`join` makes sense as a private authenticated call, however it is not clear what network it is joining (I am assuming that it is the network that it is already a non-validator of, but this feels a little bit weird to me).  Essentially, it needs the url of a node that it can forward the join request to (using broadcast).

`approve` and `leave` should definitely be using the authenticated service.

Maybe we can have a discussion about this tomorrow

File: docs/admin/validators.mdx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+```
+kwil-admin validators join [--key-file KEY-FILE] [PRIVKEY]
+```
+
+The private key of the node (prospective validator) that will sign the
+transaction is specified with either a file containing the key, or a positional
+argument containing the hexadecimal string for the key.  For example:
+
+```sh
+$ ./kwild validator join --key-file ~/secrets/validator_private_key
```
Comment (brennanjl) [inline docs/admin/validators.mdx:65]: I think this should be using `kwil-admin` instead of `kwild`

File: docs/admin/validators.mdx [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+```
+kwil-admin validators join-status JOINER
+```
+
+This command requires only the public key of the candidate validator `JOINER`
+with an active join request. For example,
+
+```
+$ kwild validator join-status "6ecaca8e9394c939a858c2c7b47acb1db26a96d7ab38bd702fa3820c5034e9d0"
```
Comment (brennanjl) [inline docs/admin/validators.mdx:45]: I think this should be using `kwil-admin` instead of `kwild`

File: docs/admin/node.mdx [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+:::warning
+The assumption is that the user of `kwil-admin` is the node
+operator. Access must not be granted to untrusted parties.
+:::
+
+Communication between `kwil-admin` and `kwild` is both encrypted and
+mutually-authenticated. Each side has their own key pair, where the node's is
+used to encrypt the communications, and the client's is used to authenticate
+itself with the node. Thus, setup involves generating client credentials, and
+setting them as an authorized RPC client on the Node.
```
Comment (brennanjl) [inline docs/admin/node.mdx:31]: "node", instead of "Node"

File: docs/admin/node.mdx [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+Use of the `node` commands requires the following:
+
+- The node's TLS (encrypted transport) certificate, specified with
+  `--authrpc-cert`. The default is "kwild.cert". This would be retrieved from
+  the node where it is located at `~/.kwild/rpc.cert` by default.
+- RPC client credentials. This is a key pair used to authenticate `kwil-admin`
+  (the client) with `kwild` (the server).  The key and certificate parts are
+  specified by `--tlskey` and `--tlscert`.
+- Adding the client certificate to the Node's allowed clients, which is locate
```
Comment (brennanjl) [inline docs/admin/node.mdx:41]: "node", instead of "Node"

File: docs/admin/node.mdx [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+### Generation of Credentials on the Node
+
+The node will generate its own TLS key and certificate, which is used encrypt
+the communications on its authenticated RPC server. As a convenience, `kwild`
+will also generate new *client* credentials on start up if the file
+`~/.kwild/clients.pem` does not exist.
+
+The corresponding `auth.{cert,key}` files may then be used by the operator to
+with th `kwil-admin node` commands for remote access to the node.
```
Comment (jchappelow) [inline docs/admin/node.mdx:62]: Yikes typos galore here, sorry

File: docs/admin/setup.mdx [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+* resetting node state (`reset-state`) and all node data files (`reset`)
+
+## Quick Setup of a Single Kwil Node
+
+The `init` command facilitates quick setup of an isolated Kwil node on a fresh
+network in which that node is the single validator.  This permits rapid
+prototyping and evaluation of Kwil DB functionality.
+
+For example, to create a node configuration for a new network in a root Kwild
+Daemon folder at `~/.kwild-new`:
```
Comment (jchappelow) [inline docs/admin/setup.mdx:24]: Oofff, need to proof these docs again

**Comment by brennanjl** [comment]
Aside from a few very minor typo things, as well as this: https://github.com/kwilteam/docs/pull/25#discussion_r1330726565 ,  I think it looks really good.

File: docs/admin/validators.mdx (hunk 7 above)
Comment (jchappelow) [reply docs/admin/validators.mdx:19]: Yeah, we should all get on the same page about the validator actions.

These general M.O. of validator commands is unchanged since they were added when the first comet bft pr was merged to main.  That is, other than `list` and `join-status`:
- node key is on the machine running the command
- kwil-admin (previously kwild utils) authors and signs one of the three transactions (join, leave, approve)
- the transaction is broadcasted on _any public RPC service_

It's exactly because this is unchanged that these commands are not subcommands of the `node` command (yet).

That all makes sense if you think of if like a wallet, which has keys, makes transactions, and broadcasts.  I believe this model would be intuitive to many people.  This approach requires having the keys on the administrative machine.  It also allows using an arbitrary RPC endpoint to broadcast transactions on behalf of any number of nodes that the operator controls, without having to touch those node machines themselves.  These would not use an authenticated RPC service for broadcast.

On the other hand, we've discussed changing to a model where kwil-admin simply makes a request to the node (current or candidate validator) that should do the transaction authoring and signing with it's own in-memory keys.  The security model here puts all the responsibility on the RPC authentication.  Yet it keeps the node keys on the node.  When this model is implemented, the kwil-admin commands would probably go under the `node` command, which is described as the remote node administration commands, where all auth RPC is performed.  This is likely to be error prone since you may easily get the wrong endpoint for the node, and end up signing a transaction with the wrong keys (thus accidentally making your validator leave, or attempt to join, etc)

There are pros and cons to both approaches.

Comment (brennanjl) [reply docs/admin/validators.mdx:19]: The security concern of having to copy private keys across machines really irks me.  As far as I know, best security practices say to generate a key on a machine and not have it leave.  I think our default should be operating under this model.

The other option is that this admin tool could directly access the private key, and then validator sub commands can only be used when running this on the same machine.

I think either of these would be fine, but I do think it's important that the expected functionality is to run a validator without ever copying the key.

Comment (jchappelow) [reply docs/admin/validators.mdx:19]: I understand your point of view, but if you consider the Ethereum validator model, it's somewhat opposite.  Their official and recommended approach is to use a special key generation tool, `deposit`, then _copy_ the generated keys to the validator client machine, and import it into the validator client.

https://docs.prylabs.network/docs/install/install-with-script#step-5-run-a-validator-using-prysm
https://launchpad.ethereum.org/en/checklist
https://github.com/ethereum/staking-deposit-cli#new-mnemonic-arguments (has `--num_validators` option to generate multiple validator keys at once)

You are correct that the key go to only one validator machine, however that is to prevent two instances of the same validator client running, which leads to slashing.

While the copy is mainly necessary because the key generation tool is supposed to be done on a secure offline machine (and you have backups), there are also validator management tools that use the keys.

For example, Prysm has their own validator wallet to manage _multiple_ validator keys: https://docs.prylabs.network/docs/wallet/deterministic
Their approach to leave/exit is almost exactly like ours at present: https://docs.prylabs.network/docs/wallet/exiting-a-validator
`prysmctl validator exit --wallet-dir=<path/to/wallet> --beacon-rpc-provider=<127.0.0.1:4000> `
They have actually transitioned from a direct validator client command to this prysmctl approach that uses a wallet with validator keys and an RPC endpoint. ("Note that this operation was previously facilitated by a command exposed by the Prysm validator client")

We should definitely get kwild to make these validator transactions internally, but I'm not entirely convinced there's no use case for signing join/leave/approve transactions in an external tool like kwil-admin, even when it's on a different machine (likely to be a secure machine that contains multiple generated validator keys, as is the case with ethereum's `deposit` key store and prysmctl's wallet).

**Comment by jchappelow** [comment]
Oh, I didn't yet trial moving this category under Kwil Daemon.  Any other opinions about it?

**Comment by brennanjl** [comment]
> Oh, I didn't yet trial moving this category under Kwil Daemon. Any other opinions about it?

I think it is good where it is.  I think it would be more confusing if it was in the Kwil Daemon section.
